internal/sync2/membuf.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package sync2
+
+import (
+	"bytes"
+	"io"
+	"sync"
+)
+
+// MemoryBuffer is a synchronized variable-sized memory buffer. Read will
+// block until new data is added or the buffer is closed. If the buffer is
+// closed, Read will continue reading as long as there is data in the buffer.
+type MemoryBuffer struct {
```

Comment:
Let's call it `MemoryPipe`... also you can actually use `filepipe` to implement it, since it relies on `ReadWriteCloserAt`. So if you implement a `type memoryWriterAt []byte` with `ReadAt`, `WriteAt` and `Close` then you can use it.